### PR TITLE
SUPESC-274: Allowed module usage without Silex (with Router plugin).

### PR DESCRIPTION
**Pull request by boykodev**
- Developer(s): @boykodev

- Ticket: https://spryker.atlassian.net/browse/SUPESC-274

- Academy PR: ACADEMY_URL_HERE

- Release Group: https://release.spryker.com/release-groups/view/3338

- PR Overview: https://release.spryker.com/release/pull-request-eco/Heidelpay/57

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Heidelpay             | minor                 |                       |

-----------------------------------------

##### Change log

Improvements

- Introduced `HeidelpayRouteProviderPlugin` to add Router support.
- Adjusted `existing-registration` molecule by removing duplication `paymentOption` form element.
- Adjusted `new-registration` molecule by removing duplication `paymentOption` form element.

Fixes
- Renamed `spy_payment_heidelpay.spy_payment_payone-fk_sales_order` database foreign-key to `spy_payment_heidelpay.spy_payment_heidelpay-fk_sales_order` in order to prevent name collision with other payment providers.

Deprecations
- Deprecated `HeidelpayClient::getQuoteFromSession()`.
- Deprecated `HeidelpayControllerProvider` to avoid `Silex` usage.
- Deprecated `InitializePlugin`.

### CODE REVIEW

File: src/SprykerEco/Client/Heidelpay/HeidelpayFactory.php [hunk 1]
```diff
@@ -31,7 +30,14 @@
 class HeidelpayFactory extends AbstractFactory
 {
     use SessionClientFactoryTrait;
-    use ZedRequestClientFactoryTrait;
+
+    /**
+     * @return \Spryker\Client\ZedRequest\ZedRequestClientInterface
+     */
+    public function getZedRequestClient()
```
Comment (asmarovydlo) [inline src/SprykerEco/Client/Heidelpay/HeidelpayFactory.php:37]:
```suggestion
    public function getZedRequestClient(): ZedRequestClientInterface
```

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    protected const HEIDELPAY_PAYMENT = 'heidelpay-payment';
+    protected const HEIDELPAY_PAYMENT_FAILED = 'heidelpay-payment-failed';
+    protected const HEIDELPAY_EASY_CREDIT_PAYMENT = 'heidelpay-easy-credit-payment';
+    protected const HEIDELPAY_EASY_CREDIT_INITIALIZE_PAYMENT = 'heidelpay-easy-credit-initialize-payment';
+    protected const HEIDELPAY_IDEAL_AUTHORIZE = 'heidelpay-ideal-authorize';
+    protected const HEIDELPAY_CREDIT_CARD_REGISTER = 'heidelpay-cc-register';
+    protected const HEIDELPAY_CREDIT_CARD_REGISTER_SUCCESS = 'heidelpay-cc-register-success';
+    protected const HEIDELPAY_DIRECT_DEBIT_REGISTER = 'heidelpay-dd-register';
+    protected const HEIDELPAY_DIRECT_DEBIT_REGISTER_SUCCESS = 'heidelpay-dd-register-success';
+    protected const HEIDELPAY_NOTIFICATION = 'heidelpay-notification';
```
Comment (asmarovydlo) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:25]: Please add `ROUTE_NAME` prefix to all route constants.

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        $routeCollection = $this->addHeidelpayCreditCardRegisterSuccessRoute($routeCollection);
+        $routeCollection = $this->addHeidelpayDirectDebitRegisterRoute($routeCollection);
+        $routeCollection = $this->addHeidelpayDirectDebitRegisterSuccessRoute($routeCollection);
+        $routeCollection = $this->addHeidelpayNotificationRoute($routeCollection);
+
+        return $routeCollection;
+    }
+
+    /**
+     * @param \Spryker\Yves\Router\Route\RouteCollection $routeCollection
```
Comment (asmarovydlo) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:49]: For each `add...Route` method please add `@uses` annotation that links used controller

Example:

```
    /**
     * @uses \SprykerShop\Yves\AgentPage\Controller\AgentController::indexAction()
     *
     * @param \Spryker\Yves\Router\Route\RouteCollection $routeCollection
     *
     * @return \Spryker\Yves\Router\Route\RouteCollection
     */
    protected function addOverviewRoute(RouteCollection $routeCollection): RouteCollection
```

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php (hunk 2 above)
Comment (asmarovydlo) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:25]: All route constants should be public.

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Yves\Heidelpay\Plugin\Router;
+
+use Spryker\Yves\Router\Plugin\RouteProvider\AbstractRouteProviderPlugin;
+use Spryker\Yves\Router\Route\RouteCollection;
```
Comment (asmarovydlo) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:11]: This requires adding spryker/router to composer dependencies.

**Review by asmarovydlo — CHANGES_REQUESTED** [review]
1. Please add:
   -  release overview link to the PR.
   - tags to the PR.
2.  Please avoid do not specific code related changelogs
 > Make Heidelpay compatible usage without Silex
 This one could be added to `Introduced HeidelpayRouteProviderPlugin`
3. Why release type is major?
4. Please follow the next pattern for PR names https://spryker.slack.com/archives/C03KVH4N8/p1597669214097700

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php (hunk 4 above)
Comment (boykodev) [reply src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:11]: also removed silex dependency

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php (hunk 3 above)
Comment (asmarovydlo) [reply src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:51]: Fixed only partially, please check comment one more time.

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+class HeidelpayRouteProviderPlugin extends AbstractRouteProviderPlugin
+{
+    public const HEIDELPAY_BUNDLE_NAME = 'Heidelpay';
+    public const ROUTE_HEIDELPAY_PAYMENT = 'heidelpay-payment';
```
Comment (asmarovydlo) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:16]: https://github.com/spryker-eco/heidelpay/pull/57#discussion_r598957456 not fixed

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+ */
+
+namespace SprykerEco\Yves\Heidelpay\Plugin\Router;
+
+use Spryker\Yves\Router\Plugin\RouteProvider\AbstractRouteProviderPlugin;
+use Spryker\Yves\Router\Route\RouteCollection;
+
+class HeidelpayRouteProviderPlugin extends AbstractRouteProviderPlugin
+{
+    public const HEIDELPAY_BUNDLE_NAME = 'Heidelpay';
```
Comment (asmarovydlo) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:15]: https://github.com/spryker-eco/heidelpay/pull/57#discussion_r598960838 - HEIDELPAY_BUNDLE_NAME is nor route constant, and should be protected.

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php (hunk 4 above)
Comment (asmarovydlo) [reply src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:11]: you've removed nothing related to Silex in the code, only deprecated.
What is the reason for dependency removal?

File: composer.json [hunk 7]
```diff
@@ -19,6 +19,7 @@
     "spryker/oms": "^7.0.0 || ^8.0.0 || ^10.0.0 || ^11.0.0",
     "spryker/propel-orm": "^1.5.0",
     "spryker/quote": "^1.0.0 || ^2.0.0",
+    "spryker/router": "^1.11",
```
Comment (asmarovydlo) [inline composer.json:22]: There is no such version like `^1.11"`
 
 ```suggestion
    "spryker/router": "^1.1.0",
```

**Review by asmarovydlo — CHANGES_REQUESTED** [review]
https://github.com/spryker-eco/heidelpay/pull/57#pullrequestreview-617829312 

Item #4 - not fixed
Item #1 - fixed only partially

Please read https://spryker.atlassian.net/wiki/spaces/CORE/pages/523730976/Release+Notes+and+Changelogs and update changelogs accordingly.

File: composer.json (hunk 7 above)
Comment (boykodev) [reply composer.json:22]: it was added automatically by composer require

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php (hunk 3 above)
Comment (boykodev) [reply src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:51]: also added action methods

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php (hunk 4 above)
Comment (boykodev) [reply src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:11]: brought silex back

File: composer.json (hunk 7 above)
Comment (asmarovydlo) [reply composer.json]: When we need to add some dependency to the module, we usually should use the lowest minor (where our code is available) of the latest major, in our case it's ^1.1.0
See https://github.com/spryker/router/releases/tag/1.1.0

**Review by asmarovydlo — APPROVED** [review]

**Review by gechetspr — COMMENTED** [review]
Code is ok, but I would rephrase `Introduced HeidelpayRouteProviderPlugin to make module work without Silex.` changelog item. This module still requires Silex as a dependency, so just specify actual purpose of this plugin

**Comment by asmarovydlo** [comment]
Please move any touched eco module to Github Actions
See an example at https://github.com/spryker/event-behavior/blob/master/.github/workflows/ci.yml

**Review by gechetspr — APPROVED** [review]

**Comment by boykodev** [comment]
Couldn't make travis tests pass. Errors of that type are not reproducible on local (all tests are green). But in travis I get this:

`Test  Business/HeidelpayFacadeAuthorizeOnRegistrationPaymentTest.php:testProcessSuccessAuthorizeOnRegistrationPayment
                                                                                                                                                        
  [Spryker\Zed\Oms\Business\Exception\StatemachineException] Could not find "6059fbdbd1a2e-process.xml". Please check your process definition location`

  https://travis-ci.org/github/spryker-eco/heidelpay/jobs/764231905

**Comment by pushokwhite** [comment]
> Couldn't make travis tests pass.
I've faced this issue and fixed it here: https://github.com/spryker-eco/heidelpay/pull/56
Please check [this](https://github.com/spryker-eco/heidelpay/pull/56/commits/e1b570b10f4a878788805554a8c6679337118363) and [this](https://github.com/spryker-eco/heidelpay/pull/56/commits/07666bad4295f13ee22dbb4b477de2db05f234b2) commits (do not look at `codecetion.yml`). If you want some explanations just ping me directly.

File: src/SprykerEco/Yves/Heidelpay/Plugin/Provider/HeidelpayControllerProvider.php [hunk 8]
```diff
@@ -10,6 +10,9 @@
 use Silex\Application;
 use SprykerShop\Yves\ShopApplication\Plugin\Provider\AbstractYvesControllerProvider;
 
+/**
+ * @deprecated Use {@link \SprykerEco\Yves\Heidelpay\Plugin\Router\HeidelpayRouteProvider} instead.
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Provider/HeidelpayControllerProvider.php:14]:
```suggestion
 * @deprecated Use {@link \SprykerEco\Yves\Heidelpay\Plugin\Router\HeidelpayRouteProviderPlugin} instead.
```

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+     *
+     * @param \Spryker\Yves\Router\Route\RouteCollection $routeCollection
+     *
+     * @return \Spryker\Yves\Router\Route\RouteCollection
+     */
+    protected function addHeidelpayPaymentRoute(RouteCollection $routeCollection): RouteCollection
+    {
+        $routeCollection->add(
+            static::ROUTE_HEIDELPAY_PAYMENT,
+            $this->buildPostRoute('/heidelpay/payment', static::HEIDELPAY_BUNDLE_NAME, 'Heidelpay', 'payment')
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:61]: that was `GET` route originally from controller provider

**Review by dmytro-sokolovskyi — COMMENTED** [review]
Could you check Travis regarding compatibility modules?

heidelpay is not compatible with the modules used in b2c-demo-shop
```
5118
Updated dependencies in module to match b2c-demo-shop
5119
Changing spryker/quote '^1.0.0 || ^2.0.0' to '^2.14.3'
5120
Changing spryker/symfony '^3.0.0' to '^3.4.0'
```


heidelpay is not compatible with the modules used in b2b-demo-shop
```
5593
Updated dependencies in module to match b2b-demo-shop
5594
Changing spryker/symfony '^3.0.0' to '^3.4.0'
```

**Review by dmytro-sokolovskyi — APPROVED** [review]

File: .travis.yml [hunk 10]
```diff
@@ -1,4 +1,6 @@
 language: php
+os: linux
```
Comment (abitskil) [inline .travis.yml:2]: * Remove the travis.
* Run the following command to fix composer.json 
* `vendor/bin/console dev:composer:update-json-files -m SprykerEco.Heidelpay`.

File: src/SprykerEco/Client/Heidelpay/HeidelpayClient.php [hunk 11]
```diff
@@ -62,12 +62,12 @@ public function getAuthorizeTransactionLogForOrder(string $orderReference): Heid
...
 
     /**
-     * @deprecated Use `getQuote()` instead.
-     *
-     * {@inheritdoc}
+     * {@inheritDoc}
      *
      * @api
      *
+     * @deprecated Use `getQuote()` instead.
```
Comment (abitskil) [inline src/SprykerEco/Client/Heidelpay/HeidelpayClient.php:69]: Use `{@link \SprykerEco\Client\Heidelpay\HeidelpayClient::getQuote()}`.

File: src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+namespace SprykerEco\Yves\Heidelpay\Plugin\Router;
+
+use Spryker\Yves\Router\Plugin\RouteProvider\AbstractRouteProviderPlugin;
+use Spryker\Yves\Router\Route\RouteCollection;
+
+class HeidelpayRouteProviderPlugin extends AbstractRouteProviderPlugin
+{
+    protected const HEIDELPAY_BUNDLE_NAME = 'Heidelpay';
+    public const ROUTE_HEIDELPAY_PAYMENT = 'heidelpay-payment';
```
Comment (abitskil) [inline src/SprykerEco/Yves/Heidelpay/Plugin/Router/HeidelpayRouteProviderPlugin.php:16]: * `ROUTE_NAME` - must be a prefix for a constants.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        strategy:
+            fail-fast: false
+            matrix:
+                include:
+                    - php: '7.4'
+
+        env:
+            APPLICATION_ENV: devtest
+            APPLICATION_STORE: DE
+            PROJECT: adyen
```
Comment (abitskil) [inline .github/workflows/ci.yml:23]: heidelpay

**Review by kkicheglovspryker — APPROVED** [review]

**Review by geronica — APPROVED** [review]

**Review by dmytro-sokolovskyi — APPROVED** [review]

**Review by abitskil — APPROVED** [review]

**Review by abitskil — APPROVED** [review]
Looks good from BE side.

**Review by FRITS-Kh — APPROVED** [review]
FE part is OK.
Please add FE changelog according to the doc
https://spryker.atlassian.net/wiki/spaces/CORE/pages/961414228/Work+in+Progress+Frontend+Change+logs

**Review by FRITS-Kh — APPROVED** [review]

File: src/SprykerEco/Zed/Heidelpay/Persistence/Propel/Schema/spy_heidelpay.schema.xml [hunk 14]
```diff
@@ -11,20 +11,20 @@
...
+        <column name="easy_credit_fee" required="false" type="INTEGER"/>
 
         <column name="date_of_birth" required="false" type="DATE"/>
-        <column name="salutation" required="false" type="VARCHAR"/>
+        <column name="salutation" required="false" type="VARCHAR" size="255"/>
 
         <column name="connector_invoice_account_info" required="false" type="LONGVARCHAR" />
 
-        <foreign-key name="spy_payment_payone-fk_sales_order" foreignTable="spy_sales_order" phpName="SpySalesOrder">
+        <foreign-key name="spy_payment_heidelpay-fk_sales_order" foreignTable="spy_sales_order" phpName="SpySalesOrder">
```
[the remark this answers is not in the document]
Comment (abitskil) [reply src/SprykerEco/Zed/Heidelpay/Persistence/Propel/Schema/spy_heidelpay.schema.xml:27]: It was decided to release it as a minor, because a foreign-key constraint is a database object as well, and thus its name needs to be unique. In the current situation, we have an intersection with the Payone payment provider.
